transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```diff
@@ -365,174 +307,124 @@ private void invokeExceptionCaught(final Throwable cause) {
+                    // Duplex handlers implements both out/in interfaces causing a scalability issue
+                    // see https://bugs.openjdk.org/browse/JDK-8180450
+                    final ChannelHandler handler = next.handler();
+                    final DefaultChannelPipeline.HeadContext headContext = next.pipeline.head;
+                    if (handler == headContext) {
+                        headContext.userEventTriggered(next, event);
+                    } else if (handler instanceof ChannelInboundHandlerAdapter) {
+                        ((ChannelInboundHandlerAdapter) handler).userEventTriggered(next, event);
+                    } else {
+                        ((ChannelInboundHandler) handler).userEventTriggered(next, event);
+                    }
+                } catch (Throwable t) {
+                    next.invokeExceptionCaught(t);
                 }
-            } catch (Throwable t) {
-                invokeExceptionCaught(t);
+            } else {
+                next.fireUserEventTriggered(event);
             }
         } else {
-            fireUserEventTriggered(event);
+            next.executor().execute(() -> fireUserEventTriggered(event));
         }
+        return this;
     }
 
     @Override
     public ChannelHandlerContext fireChannelRead(final Object msg) {
-        invokeChannelRead(findContextInbound(MASK_CHANNEL_READ), msg);
-        return this;
-    }
-
-    static void invokeChannelRead(final AbstractChannelHandlerContext next, Object msg) {
-        final Object m = next.pipeline.touch(ObjectUtil.checkNotNull(msg, "msg"), next);
-        EventExecutor executor = next.executor();
-        if (executor.inEventLoop()) {
-            next.invokeChannelRead(m);
-        } else {
-            executor.execute(new Runnable() {
-                @Override
-                public void run() {
-                    next.invokeChannelRead(m);
-                }
-            });
-        }
-    }
-
-    private void invokeChannelRead(Object msg) {
-        if (invokeHandler()) {
-            try {
-                // DON'T CHANGE
-                // Duplex handlers implements both out/in interfaces causing a scalability issue
-                // see https://bugs.openjdk.org/browse/JDK-8180450
-                final ChannelHandler handler = handler();
-                final DefaultChannelPipeline.HeadContext headContext = pipeline.head;
-                if (handler == headContext) {
-                    headContext.channelRead(this, msg);
-                } else if (handler instanceof ChannelDuplexHandler) {
-                    ((ChannelDuplexHandler) handler).channelRead(this, msg);
-                } else {
-                    ((ChannelInboundHandler) handler).channelRead(this, msg);
+        AbstractChannelHandlerContext next = findContextInbound(MASK_CHANNEL_READ);
+        if (next.executor().inEventLoop()) {
+            final Object m = next.pipeline.touch(msg, next);
```

Comment:
this is similar to what I've said at https://github.com/netty/netty/pull/14705/files#r1924693265